### PR TITLE
Remove explicit prolongator construction

### PR DESCRIPTION
**Pull request by aprokop**
This PR adds support for `apply()` in transpose mode, and removes prolongator from the hierarchy, replacing it by action of restrictor in transpose mode.

As we don't have a proper support for transpose action in device sparse matrices, I chose to workaround that by using the implemented `transpose()` call of `CudaMatrixOperator`. That class now caches the matrix transpose when needed. I am not sure how to avoid `mutable` class variables, though, and open to suggestions.

### CODE REVIEW

File: include/mfmg/common/level.hpp [hunk 1]
```diff
@@ -58,20 +53,22 @@ class Level
...
   {
     _smoother = s;
   }
 
   void set_solver(std::shared_ptr<Solver<vector_type> const> s) { _solver = s; }
 
+  std::shared_ptr<vector_type> build_vector() const
+  {
+    auto a = get_operator();
+    assert(a);
```
Comment (Rombur) [inline include/mfmg/common/level.hpp]: We have our own assert

File: include/mfmg/common/operator.hpp [hunk 2]
```diff
@@ -16,6 +16,12 @@
 
 namespace mfmg
 {
+enum OperatorMode
```
Comment (Rombur) [inline include/mfmg/common/operator.hpp]: Use `enum class`

File: include/mfmg/cuda/cuda_matrix_operator.cuh [hunk 3]
```diff
@@ -50,7 +51,8 @@ public:
   std::shared_ptr<SparseMatrixDevice<value_type>> get_matrix() const;
 
 private:
-  std::shared_ptr<SparseMatrixDevice<value_type>> _matrix;
+  mutable std::shared_ptr<SparseMatrixDevice<value_type>> _matrix;
+  mutable std::shared_ptr<SparseMatrixDevice<value_type>> _transposed_matrix;
```
Comment (Rombur) [inline include/mfmg/cuda/cuda_matrix_operator.cuh]: Why do you need to store both matrices?

File: include/mfmg/common/operator.hpp (hunk 2 above)
Comment (aprokop) [reply include/mfmg/common/operator.hpp]: Thank you, was not aware of this.

File: include/mfmg/cuda/cuda_matrix_operator.cuh (hunk 3 above)
Comment (aprokop) [reply include/mfmg/cuda/cuda_matrix_operator.cuh]: The goal was to try to minimize the number of transpose calculations. Transpose actually appears in multiple places (through `apply(TRANS)` and `transpose()`, for example), so by storing it we essentially cache the calculation.

**Pull request merged**